Agency Dashboard: Add Request to Check Sites Connection Health

wilmersondasilva opened the pull request:
Currently, the connection health status is populated in the site profiles index during an audit, which means that there might be a significant delay between the time site went down and the index update. More context can be found pedEI5-2s-p2 here.

In WordPress.com, we use the `/test-connection` endpoint that makes an API request to the remote site to verify the connection's health. 

The API was changed so we can use this endpoint also on Calypso green.

**Backend task**: 1202619025189113-as-1203022769710885

#### Proposed Changes

- In the Calypso dashboard, the connection status of each site will be fetched.
- If the connection status of the site is not healthy, an error message will appear to the user.

#### Testing Instructions

- Make sure to have a partner account.
- Create an ephemeral site
- Connect Jetpack
- Ensure the site is now visible in the [Agency Dashboard](https://calypso.live/?image=registry.a8c.com/calypso/app:build-45797&env=jetpack)
- Delete the ephemeral site
- Visit the site to ensure it is gone
- Refresh the [Agency Dashboard](https://calypso.live/?image=registry.a8c.com/calypso/app:build-45797&env=jetpack) page
- Verify that there is now an indication about the unhealthy site connection

#### Screenshots

![image](https://user-images.githubusercontent.com/6406071/194975504-f0915ca4-eac8-4ab5-8b1d-c89ce77e88be.png)


#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to 1202619025189113-as-1203022769710885

wilmersondasilva committed:
Agency Dashboard: create use-fetch-test-connection hook to fetch connection status

This hook will be used to check the connection status of a site

wilmersondasilva committed:
Agency Dashboard: create site-table-row component

This is a refactor of the site-table component, which part of it was turned now into this new component site-table-row

wilmersondasilva committed:
Agency Dashboard: test the sites connections in the site-card component

Add the usage of the useFetchTestConnection hook to check the sites connection status

wilmersondasilva committed:
Agency Dashboard: fix bug in the use-fetch-test-connection hook

The hook was using wpcomJetpackLicensing to fetch data, but since the API used here requires a different authentication mechanism, I need to use wpcom instead.

wilmersondasilva committed:
Agency Dashboard: update site-table and site-table-row component unit tests

Update the unit tests to reflect the latest changes in the components

wilmersondasilva committed:
Agency Dashboard: decrease amount of HTTP requests made to check the sites connection status

There are two components to show the list of sites. One for mobile and the other for desktop. They were shown/hidden using CSS. This PR moves this conditional out from CSS and implements it using React.

rcanepa commented:
I did a quick manual test and noticed a discrepancy between the number of sites I see on production versus the live branch associated with this PR.

<table>
<tr>
<th>
PR
</th>
<th>
Production
</th>
</tr>
<tr>
<td>
<img src="https://user-images.githubusercontent.com/3418513/195194176-9d356789-d472-41ab-9e75-502a78fabc6c.png" />
</td>
<td>
<img src="https://user-images.githubusercontent.com/3418513/195194237-9ee65e35-487c-47a6-be0c-21bed597aec3.png" />
</td>
</tr>
</table>

Additionally, there are two unit tests failing at this moment.

vitozev commented:
> I did a quick manual test and noticed a discrepancy between the number of sites I see on production versus the live branch associated with this PR.

@rcanepa, now when you say this, I remember that I saw this issue on the horizon environment back in time. It's not caused by this PR, and it can be reproduced in other PRs:

- https://github.com/Automattic/wp-calypso/pull/68922

I've added Asana card: 1202619025189113-as-1203147992256213

wilmersondasilva committed:
Agency Dashboard: update site-card and site-content components unit tests

These unit tests were breking due the latest changes in their components

wilmersondasilva commented:
@vitozev, thank you for pointing out this issue. I haven't touched the sidebar code, so it makes sense it's not something related to the PR.

@rcanepa, I just updated the two failing unit tests.

yashwin commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-content/index.tsx:
```diff
@@ -24,6 +24,7 @@ interface Props {
 
 export default function SiteContent( { data, isLoading, currentPage, isFavoritesTab }: Props ) {
 	const isMobile = useMobileBreakpoint();
+	const isDesktop = useDesktopBreakpoint();
```
How about we can use the `isMobile` condition to display the mobile view and fall back to a large screen view if not mobile?

wilmersondasilva replied:
`isMobile` is true for screens smaller than **480px**, while `isDesktop` is true for screens bigger than **960px**.

yashwin commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-error-content.tsx:
```diff
@@ -12,11 +12,14 @@ export default function SiteErrorContent( { siteUrl }: { siteUrl: string } ) {
-			<span className="sites-overview__error-message sites-overview__error-message-large-screen">
+			<span
+				className="sites-overview__error-message sites-overview__error-message-large-screen"
+				data-testid="site-error-message"
```
How about we get rid of the `data-testid` and use `screen.getByText` instead to get the element since it is close to what the user sees on the screen?

wilmersondasilva replied:
@yashwin I pushed the suggested changes in the unit tests.

yashwin commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-error-content.tsx:
```diff
@@ -30,6 +33,7 @@ export default function SiteErrorContent( { siteUrl }: { siteUrl: string } ) {
 				onClick={ handleClickFixNow }
 				className="sites-overview__error-message-link"
 				href={ `/settings/disconnect-site/${ siteUrl }?type=down` }
+				data-testid="site-error-message-link"
```
Same here, we could use `screen.getByText( /fix now/i )`

yashwin commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-error-content.tsx:
```diff
@@ -12,11 +12,14 @@ export default function SiteErrorContent( { siteUrl }: { siteUrl: string } ) {
 
 	return (
-		<div className="sites-overview__error-container">
+		<div className="sites-overview__error-container" data-testid="site-error-container">
```
I don't think we are using `site-error-container` anywhere, which we can eliminate.

yashwin commented:
I tested the changes. It works as expected. Thanks for making these changes and adding the test cases, @wilmersondasilva :) 

I found a tiny issue here. The UI didn't show the `Jetpack is unable to connect to {site}` error message for a while, even though the API had returned false.

<img width="746" alt="Screenshot 2022-10-12 at 1 02 00 PM" src="https://user-images.githubusercontent.com/10586875/195279121-c291bf6c-48b0-4f4b-afd5-9b55eadc9a60.png">

wilmersondasilva committed:
Agency Dashboard: update the site-table-row component unit tests

Replace the usage of getByTestId by getByText and remove unnecessaries data-testid attributes

wilmersondasilva committed:
Agency Dashboard: fix bug in site-table-row and site-card

Both components were using useMemo where it shouldn't be used.

wilmersondasilva commented:
@yashwin, thanks for reviewing it and for bringing up this issue. I just fixed it.

yashwin reviewed (approved):
LGTM now, thanks for making the changes. I see there is one test case failing, you might have to re-authorize circle ci to fix it.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7578031

Thank you @wilmersondasilva for including a screenshot in the description! This is really helpful for our translators.

vitozev commented:
@wilmersondasilva, I know I'm a little late to the party, but maybe we should add some caching (1 minute?) to avoid triggering the HTTP requests often. This might be helpful if users navigate back and forth between pages in the dashboard.

In addition, we should probably disable the [window focus refetching](https://react-query-v3.tanstack.com/guides/window-focus-refetching), as I don't think it brings much value in that specific case.

What do you think?

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7578031).
